### PR TITLE
feat: response authentication (certificate verification, root key lookup, delegations)

### PR DESCRIPTION
**Pull request by ericswanson-dfinity**
Implements from doc:
- Checking BLS signature
- Fetching root key from /api/v1/status for dev instances
- Hard-coding Sodium key (**still need the actual key**).  (Note doc says this would be the Mercury key)
- Support for delegations 

See also companion PR [sdk #1197](https://github.com/dfinity/sdk/pull/1197)

### CODE REVIEW

File: ic-agent/src/agent/mod.rs [hunk 1]
```diff
@@ -28,11 +28,22 @@ use reqwest::Method;
...
 
-const DOMAIN_SEPARATOR: &[u8; 11] = b"\x0Aic-request";
+const IC_REQUEST_DOMAIN_SEPARATOR: &[u8; 11] = b"\x0Aic-request";
+const IC_STATE_ROOT_DOMAIN_SEPARATOR: &[u8; 14] = b"\x0Dic-state-root";
+const DER_PREFIX: &[u8; 37] = b"\x30\x81\x82\x30\x1d\x06\x0d\x2b\x06\x01\x04\x01\x82\xdc\x7c\x05\x03\x01\x02\x01\x06\x0c\x2b\x06\x01\x04\x01\x82\xdc\x7c\x05\x03\x02\x01\x03\x61\x00";
+const KEY_LENGTH: usize = 96;
+
+static INIT_BLS: Once = Once::new();
+
+// todo: do not merge until this is the actual Sodium key
```
Comment (p-shahi) [inline ic-agent/src/agent/mod.rs:45]: ?

File: ic-agent/src/agent/mod.rs [hunk 2]
```diff
@@ -428,6 +495,37 @@ impl Agent {
     }
 }
 
+fn initialize_bls() -> Result<(), AgentError> {
```
Comment (p-shahi) [inline ic-agent/src/agent/mod.rs:498]: This file is getting big, does it make sense to move this, `extract_der`, and the lookup functions to a different file?

File: ic-agent/src/agent/agent_error.rs [hunk 3]
```diff
@@ -78,6 +78,23 @@ pub enum AgentError {
...
+    #[error(
+        r#"BLS DER-encoded public key must be ${expected} bytes long, but is {actual} bytes long."#
+    )]
+    DerKeyLengthMismatch { expected: usize, actual: usize },
+
+    #[error("BLS DER-encoded public key is invalid. Expected the following prefix: ${expected:?}, but got ${actual:?}")]
+    DerPrefixMismatch { expected: Vec<u8>, actual: Vec<u8> },
+
+    #[error("The status response did not contain a root key")]
+    NoRootKeyInStatus(),
```
Comment (p-shahi) [inline ic-agent/src/agent/agent_error.rs:94]: should we also print the `status` here (may be useful)

File: ic-agent/src/agent/mod.rs [hunk 4]
```diff
@@ -136,9 +151,25 @@ impl Agent {
             ingress_expiry_duration: config
                 .ingress_expiry_duration
                 .unwrap_or_else(|| Duration::from_secs(300)),
+            root_key: RwLock::new(DEFAULT_ROOT_KEY.to_vec()),
         })
     }
 
+    /// Fetch the root key from the status endpoint.
+    /// It is not necessary to call this when communicating with "the" Internet Computer.
```
Comment (p-shahi) [inline ic-agent/src/agent/mod.rs:159]: nit: maybe reword to affirm the positive "Only an agent communicating with a local/custom instance of the Internet Computer should call this method. It is not necessary to call this when communicating with the DFINITY foundation managed Internet Computer."

File: ic-agent/src/agent/mod.rs [hunk 5]
```diff
@@ -136,9 +151,25 @@ impl Agent {
...
+    pub async fn fetch_root_key(&self) -> Result<(), AgentError> {
+        let status = self.status().await?;
+        let root_key = status.root_key.ok_or_else(AgentError::NoRootKeyInStatus)?;
+        let mut write_guard = self.root_key.write().unwrap();
+        *write_guard = root_key;
```
Comment (p-shahi) [inline ic-agent/src/agent/mod.rs:164]:
```suggestion
        if let Ok(mut write_guard) = self.root_key.write() {
            *write_guard = root_key;
        }
```

File: ic-agent/src/agent/mod.rs [hunk 6]
```diff
@@ -136,9 +151,25 @@ impl Agent {
...
+        Ok(())
+    }
+
+    fn read_root_key(&self) -> Result<Vec<u8>, AgentError> {
+        let root_key = self.root_key.read().unwrap().clone();
+        Ok(root_key)
```
Comment (p-shahi) [inline ic-agent/src/agent/mod.rs:170]:
```suggestion
        if let Ok(read_lock) = self.root_key.read() {
            let root_key = *read_lock;
            Ok(root_key)
        } else {
            Err(AgentError::CouldntReadRootKey)
        }
```

File: ic-agent/src/agent/mod.rs (hunk 5 above)
Comment (p-shahi) [reply ic-agent/src/agent/mod.rs:164]: getting rid of the unwraps

File: ic-agent/src/agent/mod.rs (hunk 6 above)
Comment (p-shahi) [reply ic-agent/src/agent/mod.rs:170]: getting rid of unwraps and adding some new AgentError

File: ic-agent/src/agent/mod.rs (hunk 4 above)
Comment (ericswanson-dfinity) [reply ic-agent/src/agent/mod.rs:159]: Thanks!

File: ic-agent/src/agent/mod.rs (hunk 5 above)
Comment (ericswanson-dfinity) [reply ic-agent/src/agent/mod.rs:164]: Ok, I applied these changes.  The reason I had them as `unwrap()` is because the RwLock poisoning condition that would cause them to fail can't happen in the code as is, because the writer can't panic.

File: ic-agent/src/agent/mod.rs (hunk 2 above)
Comment (ericswanson-dfinity) [reply ic-agent/src/agent/mod.rs:498]: Good idea!

**Review by p-shahi — APPROVED** [review]
